panic: runtime error: invalid memory address or nil pointer dereference -- kubelet

Here is the command while running kubelet

```
/usr/local/bin/kubelet --api-servers=https://my-kubernetes-master:443 --cluster-dns=10.0.0.10 --cluster-domain=cluster.local --pod-cidr=10.0.0.0/24 --kubeconfig=/root/kubeconfig.cfg --v=9 1>&2 2> panic.log
```

Here is my OS info, and K8S version is <code>1.4.5</code>

```
cat /etc/centos-release
CentOS Linux release 7.2.1511 (Core)
uname -a
Linux tivsun01.cn.ibm.com 3.10.0-327.36.3.el7.x86_64 #1 SMP Mon Oct 24 16:09:20 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
kubelet --version
Kubernetes v1.4.5
```


```
Flag --api-servers has been deprecated, Use --kubeconfig instead. Will be removed in a future version.
I1213 14:37:07.544761   10533 loader.go:354] Config loaded from file /root/kubeconfig.cfg
I1213 14:37:07.545479   10533 docker.go:375] Connecting to docker on unix:///var/run/docker.sock
I1213 14:37:07.545508   10533 docker.go:395] Start docker client with request timeout=2m0s
E1213 14:37:07.545592   10533 cni.go:163] error updating cni config: No networks found in /etc/cni/net.d
I1213 14:37:07.555356   10533 manager.go:143] cAdvisor running in container: "/"
W1213 14:37:07.728420   10533 manager.go:151] unable to connect to Rkt api service: rkt: cannot tcp Dial rkt api service: dial tcp [::1]:15441: getsockopt: connection refused
I1213 14:37:07.893210   10533 fs.go:117] Filesystem partitions: map[/dev/mapper/vg_0c64d3bb3d0038719cf16c98ff57d7c7-brick_08f1923d65b741a452e5a8e23c70b043:{mountpoint:/var/lib/heketi/mounts/vg_0c64d3bb3d0038719cf16c98ff57d7c7/brick_08f1923d65b741a452e5a8e23c70b043 major:253 minor:9 fsType:xfs blockSize:0}
... ... ... ...
I1213 14:37:07.900561   10533 manager.go:204] Version: {KernelVersion:3.10.0-327.36.3.el7.x86_64 ContainerOsVersion:CentOS Linux 7 (Core) DockerVersion:1.10.0 CadvisorVersion: CadvisorRevision:}
I1213 14:37:07.903359   10533 oom_linux.go:69] attempting to set "/proc/self/oom_score_adj" to "-999"
W1213 14:37:07.903555   10533 server.go:613] No api server defined - no events will be sent to API server.
I1213 14:37:07.903575   10533 server.go:644] Using root directory: /var/lib/kubelet
W1213 14:37:07.907011   10533 kubelet_network.go:71] Hairpin mode set to "promiscuous-bridge" but configureCBR0 is false, falling back to "hairpin-veth"
I1213 14:37:07.907057   10533 kubelet.go:513] Hairpin mode set to "hairpin-veth"
I1213 14:37:08.084206   10533 docker_manager.go:242] Setting dockerRoot to /var/lib/docker
I1213 14:37:08.084500   10533 plugins.go:56] Registering credential provider: .dockercfg
I1213 14:37:08.084748   10533 kubelet_network.go:306] Setting Pod CIDR:  -> 10.0.0.0/24
I1213 14:37:08.085305   10533 plugins.go:355] Loaded volume plugin "kubernetes.io/aws-ebs"
I1213 14:37:08.085385   10533 plugins.go:355] Loaded volume plugin "kubernetes.io/empty-dir"
I1213 14:37:08.085440   10533 plugins.go:355] Loaded volume plugin "kubernetes.io/gce-pd"
... ... ... ...
I1213 14:37:08.088186   10533 server.go:714] Started kubelet v1.4.5
W1213 14:37:08.088297   10533 kubelet.go:1160] No api server defined - no node status update will be sent.
E1213 14:37:08.088297   10533 kubelet.go:1091] Image garbage collection failed: unable to find data for container /
I1213 14:37:08.088346   10533 server.go:118] Starting to listen on 0.0.0.0:10250
I1213 14:37:08.088407   10533 server.go:608] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"my-kubelet-host", UID:"my-kubelet-host", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'Starting' Starting kubelet.
I1213 14:37:08.088484   10533 kubelet_node_status.go:203] Setting node annotation to enable volume controller attach/detach
I1213 14:37:08.088411   10533 server.go:135] Starting to listen read-only on 0.0.0.0:10255
I1213 14:37:08.088385   10533 config.go:98] Looking for [], have seen map[]
I1213 14:37:08.141506   10533 image_gc_manager.go:171] Pod default/kafka-consumer-566n0, container kafka-consumer uses image atlas/kafka:ACT(sha256:bf8f4bf32893f7e3bc6e89558bb5752b7e90ea0097e0021227b631517a894eae)
I1213 14:37:08.141558   10533 image_gc_manager.go:171] Pod default/kafka-consumer-566n0, container POD uses image gcr.io/google_containers/pause-amd64:3.0(sha256:f1a957c92b0511d64a0c7933a77f44079c942963084e11ba2c6a43d5649c4481)
I1213 14:37:08.141566   10533 image_gc_manager.go:171] Pod default/kafka-consumer-vpiy1, container kafka-consumer uses image atlas/kafka:ACT(sha256:bf8f4bf32893f7e3bc6e89558bb5752b7e90ea0097e0021227b631517a894eae)
I1213 14:37:08.141572   10533 image_gc_manager.go:171] Pod default/kafka-consumer-vpiy1, container POD uses image gcr.io/google_containers/pause-amd64:3.0(sha256:f1a957c92b0511d64a0c7933a77f44079c942963084e11ba2c6a43d5649c4481)
I1213 14:37:08.141578   10533 image_gc_manager.go:171] Pod default/kafka-2, container POD uses image gcr.io/google_containers/pause-amd64:3.0(sha256:f1a957c92b0511d64a0c7933a77f44079c942963084e11ba2c6a43d5649c4481)
I1213 14:37:08.141586   10533 image_gc_manager.go:182] Adding image ID sha256:e033df15625593d113e96be688b9e268c98ced77c9908610bb329bd1dd0e0853 to currentImages
I1213 14:37:08.141593   10533 image_gc_manager.go:187] Image ID sha256:e033df15625593d113e96be688b9e268c98ced77c9908610bb329bd1dd0e0853 is new
I1213 14:37:08.141600   10533 image_gc_manager.go:199] Image ID sha256:e033df15625593d113e96be688b9e268c98ced77c9908610bb329bd1dd0e0853 has size 79229838
I1213 14:37:08.141609   10533 image_gc_manager.go:182] Adding image ID sha256:36216e2ba335835396634b23a23e1820e5d3068c484015f65133b2383fdbe36c to currentImages
... ... ... ...
I1213 14:37:08.142152   10533 image_gc_manager.go:199] Image ID sha256:bf595365a5588ec1bae3e9dc9efde13672f7b7561b6cc4514a82f07be9d01ca6 has size 241656
E1213 14:37:08.208733   10533 kubelet.go:2124] Failed to check if disk space is available for the runtime: failed to get fs info for "runtime": unable to find data for container /
E1213 14:37:08.208819   10533 kubelet.go:2132] Failed to check if disk space is available on the root partition: failed to get fs info for "root": unable to find data for container /
I1213 14:37:08.208838   10533 kubelet_node_status.go:356] Recording NodeHasSufficientDisk event message for node my-kubelet-host
I1213 14:37:08.208922   10533 kubelet_node_status.go:356] Recording NodeHasSufficientMemory event message for node my-kubelet-host
I1213 14:37:08.208952   10533 kubelet_node_status.go:356] Recording NodeHasNoDiskPressure event message for node my-kubelet-host
I1213 14:37:08.209140   10533 server.go:608] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"my-kubelet-host", UID:"my-kubelet-host", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeHasSufficientDisk' Node my-kubelet-host status is now: NodeHasSufficientDisk
I1213 14:37:08.209247   10533 server.go:608] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"my-kubelet-host", UID:"my-kubelet-host", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeHasSufficientMemory' Node my-kubelet-host status is now: NodeHasSufficientMemory
I1213 14:37:08.209331   10533 server.go:608] Event(api.ObjectReference{Kind:"Node", Namespace:"", Name:"my-kubelet-host", UID:"my-kubelet-host", APIVersion:"", ResourceVersion:"", FieldPath:""}): type: 'Normal' reason: 'NodeHasNoDiskPressure' Node my-kubelet-host status is now: NodeHasNoDiskPressure
I1213 14:37:08.211429   10533 fs_resource_analyzer.go:66] Starting FS ResourceAnalyzer
I1213 14:37:08.211483   10533 status_manager.go:125] Kubernetes client is nil, not starting status manager.
I1213 14:37:08.211500   10533 kubelet.go:2226] Starting kubelet main sync loop.
I1213 14:37:08.211526   10533 kubelet.go:2237] skipping pod synchronization - [network state unknown container runtime is down]
I1213 14:37:08.211563   10533 generic.go:177] GenericPLEG: Relisting
I1213 14:37:08.211591   10533 container_manager_linux.go:625] attempting to apply oom_score_adj of -999 to pid 13791
I1213 14:37:08.211609   10533 oom_linux.go:69] attempting to set "/proc/13791/oom_score_adj" to "-999"
W1213 14:37:08.211715   10533 container_manager_linux.go:667] CPUAccounting not enabled for pid: 13791
W1213 14:37:08.211724   10533 container_manager_linux.go:670] MemoryAccounting not enabled for pid: 13791
I1213 14:37:08.211718   10533 volume_manager.go:237] The desired_state_of_world populator starts
I1213 14:37:08.211746   10533 volume_manager.go:239] Starting Kubelet Volume Manager
I1213 14:37:08.211730   10533 container_manager_linux.go:371] Discovered runtime cgroups name: /system.slice/docker.service
I1213 14:37:08.211772   10533 iptables.go:362] running iptables -N [KUBE-MARK-DROP -t nat]
I1213 14:37:08.211858   10533 config.go:98] Looking for [], have seen map[]
I1213 14:37:08.211888   10533 container_manager_linux.go:625] attempting to apply oom_score_adj of -999 to pid 10533
I1213 14:37:08.211919   10533 oom_linux.go:69] attempting to set "/proc/10533/oom_score_adj" to "-999"
I1213 14:37:08.211895   10533 reconciler.go:146] Sources are all ready, starting reconstruct state function
W1213 14:37:08.212947   10533 container_manager_linux.go:667] CPUAccounting not enabled for pid: 10533
W1213 14:37:08.212985   10533 container_manager_linux.go:670] MemoryAccounting not enabled for pid: 10533
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x48 pc=0x79245f]

goroutine 176 [running]:
panic(0x2fe1980, 0xc8200120d0)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/volume/glusterfs.(*glusterfsPlugin).NewMounter(0xc8202ae090, 0xc8204e5720, 0xc8209f4280, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go:106 +0x11f
k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.(*reconciler).reconstructVolume(0xc820098580, 0xc8203af216, 0x24, 0xc8204de44b, 0x28, 0xc8204de630, 0x83, 0xc8204e5420, 0x17, 0xc820490ae0, ...)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler/reconciler.go:517 +0x399
k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.(*reconciler).syncStates(0xc820098580, 0xc8204e4f60, 0x15)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler/reconciler.go:430 +0x354
k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.(*reconciler).sync(0xc820098580)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler/reconciler.go:384 +0x68
k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.(*reconciler).reconciliationLoopFunc.func1()
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler/reconciler.go:147 +0xf4
k8s.io/kubernetes/pkg/util/wait.JitterUntil.func1(0xc820972f00)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:84 +0x19
k8s.io/kubernetes/pkg/util/wait.JitterUntil(0xc820972f00, 0x5f5e100, 0x0, 0xc820972f01, 0xc820132ea0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:85 +0xb4
k8s.io/kubernetes/pkg/util/wait.Until(0xc820972f00, 0x5f5e100, 0xc820132ea0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/util/wait/wait.go:47 +0x43
k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.(*reconciler).Run(0xc820098580, 0x7f921fa7a428, 0xc8202669c0, 0xc820132ea0)
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler/reconciler.go:133 +0x5b
created by k8s.io/kubernetes/pkg/kubelet/volumemanager.(*volumeManager).Run
        /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/volumemanager/volume_manager.go:240 +0x160

```

Here is my kubeconfig.cfg

```
apiVersion: v1
clusters:
- cluster:
    certificate-authority: /srv/kubernetes/ca.crt
    server: https://my-kubernetes-master:443
  name: cdl
contexts:
- context:
    cluster: cdl
    namespace: default
    user: blue-user
  name: blue-context
- context:
    cluster: cdl
    namespace: default
    user: green-user
  name: green-context
- context:
    cluster: cdl
    namespace: default
    user: white-user
  name: white-context
current-context: green-context
kind: Config
preferences: {}
users:
- name: blue-user
  user:
    token: /srv/kubernetes/token.csv
- name: green-user
  user:
    client-certificate-data: REDACTED
    client-key-data: REDACTED
    token: /srv/kubernetes/token.csv
- name: white-user
  user:
    password: my-password
    username: kube
    ```

I don't think it very much related to Glusterfs itself. I tried some combination and here are what I found:

1. When I use GlasterFS without --kubeconfig, it works fine.  So I have GlusterFS, Heketi, Zookeeper Clusters and Kafka PetSets working in my environment. 
2. When I configure a fresh new kube-apiserver, and run command kubelet, it crash either. 

2.1 Here you can find a fresh new kube-apiserver
```
kubectl get pvc,pv,storageclass,svc,po,all
NAME         CLUSTER-IP   EXTERNAL-IP   PORT(S)   AGE
kubernetes   10.0.0.1     <none>        443/TCP   5m
```
2.2 Then just kubelet without any parameter
```
/usr/local/bin/kubelet
...
...
E1220 14:42:17.586739   20509 kubelet.go:2124] Failed to check if disk space is available for the runtime: failed to get fs info for "runtime": unable to find data for container /
E1220 14:42:17.586827   20509 kubelet.go:2132] Failed to check if disk space is available on the root partition: failed to get fs info for "root": unable to find data for container /
I1220 14:42:17.588936   20509 fs_resource_analyzer.go:66] Starting FS ResourceAnalyzer
I1220 14:42:17.589011   20509 status_manager.go:125] Kubernetes client is nil, not starting status manager.
I1220 14:42:17.589060   20509 kubelet.go:2226] Starting kubelet main sync loop.
I1220 14:42:17.589119   20509 kubelet.go:2237] skipping pod synchronization - [network state unknown container runtime is down]
W1220 14:42:17.589170   20509 container_manager_linux.go:667] CPUAccounting not enabled for pid: 19093
W1220 14:42:17.589181   20509 container_manager_linux.go:670] MemoryAccounting not enabled for pid: 19093
I1220 14:42:17.589198   20509 volume_manager.go:239] Starting Kubelet Volume Manager
W1220 14:42:17.589249   20509 container_manager_linux.go:667] CPUAccounting not enabled for pid: 20509
W1220 14:42:17.589257   20509 container_manager_linux.go:670] MemoryAccounting not enabled for pid: 20509
panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xb code=0x1 addr=0x48 pc=0x79245f]

goroutine 157 [running]:
panic(0x2fe1980, 0xc82000a0f0)
	/usr/local/go/src/runtime/panic.go:481 +0x3e6
k8s.io/kubernetes/pkg/volume/glusterfs.(*glusterfsPlugin).NewMounter(0xc8204da000, 0xc82097e5c0, 0xc8209be780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/volume/glusterfs/glusterfs.go:106 +0x11f
k8s.io/kubernetes/pkg/kubelet/volumemanager/reconciler.(*reconciler).reconstructVolume(0xc8208080b0, 0xc820905356, 0x24, 0xc8201d6e6b, 0x28, 0xc8201d6fc0, 0x83, 0xc82097e400, 0x17, 0xc8204dde60, ...)
```

2.3 Then if I give --api-servers=https://my.host.com:443 , it works again. I just cannot get CA cert been verified

```
/usr/local/bin/kubelet --api-servers=https://9.123.102.90:443 
Flag --api-servers has been deprecated, Use --kubeconfig instead. Will be removed in a future version.
W1220 14:46:31.720036   20678 server.go:549] Could not load kubeconfig file /var/lib/kubelet/kubeconfig: stat /var/lib/kubelet/kubeconfig: no such file or directory. Using default client config instead.
I1220 14:46:31.720538   20678 docker.go:375] Connecting to docker on unix:///var/run/docker.sock
I1220 14:46:31.720561   20678 docker.go:395] Start docker client with request timeout=2m0s
E1220 14:46:31.720693   20678 cni.go:163] error updating cni config: No networks found in /etc/cni/net.d
I1220 14:46:31.726962   20678 manager.go:143] cAdvisor running in container: "/"
W1220 14:46:31.917614   20678 manager.go:151] unable to connect to Rkt api service: rkt: cannot tcp Dial rkt api service: dial tcp [::1]:15441: getsockopt: connection refused
I1220 14:46:32.089246   20678 fs.go:117] Filesystem partitions: map[/dev/mapper/vg_193739e537b2454ce37888fcd7ddb1a0-brick_db38d7559afe254824cb3bc415f1e5e5:{mountpoint:/var/lib/heketi/mounts/vg_193739e537b2454ce37888fcd7ddb1a0/brick_db38d7559afe254824cb3bc415f1e5e5 major:253 minor:10 fsType:xfs blockSize:0} /dev/mapper/vg_193739e537b2454ce37888fcd7ddb1a0-brick_4cefda9fe7d81f7f3f0a09d4711b4d48:{mountpoint:/var/lib/heketi/mounts/vg_193739e537b2454ce37888fcd7ddb1a0/brick_4cefda9fe7d81f7f3f0a09d4711b4d48 major:253 minor:15 fsType:xfs blockSize:0} /dev/mapper/vg_193739e537b2454ce37888fcd7ddb1a0-brick_5fec261416185057b6f16330d8a322f2:{mountpoint:/var/lib/heketi/mounts/vg_193739e537b2454ce37888fcd7ddb1a0/brick_5fec261416185057b6f16330d8a322f2 major:253 minor:21 fsType:xfs blockSize:0} /dev/sda2:{mountpoint:/ major:8 minor:2 fsType:xfs blockSize:0} /dev/sda1:{mountpoint:/boot major:8 minor:1 fsType:ext4 blockSize:0}]
```

--- a/pkg/volume/glusterfs/glusterfs.go
+++ b/pkg/volume/glusterfs/glusterfs.go
@@ -129,7 +129,11 @@ func (plugin *glusterfsPlugin) NewMounter(spec *volume.Spec, pod *v1.Pod, _ volu
 	ep_name := source.EndpointsName
 	// PVC/POD is in same ns.
 	ns := pod.Namespace
-	ep, err := plugin.host.GetKubeClient().Core().Endpoints(ns).Get(ep_name, metav1.GetOptions{})
+	kubeClient := plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return nil, fmt.Errorf("glusterfs: failed to get kube client to initialize mounter")
+	}
+	ep, err := kubeClient.Core().Endpoints(ns).Get(ep_name, metav1.GetOptions{})
 	if err != nil {
 		glog.Errorf("glusterfs: failed to get endpoints %s[%v]", ep_name, err)
 		return nil, err
@@ -434,7 +438,11 @@ func (d *glusterfsVolumeDeleter) GetPath() string {
 // in a given storage class, and mark them in the table.
 //
 func (p *glusterfsPlugin) collectGids(className string, gidTable *MinMaxAllocator) error {
-	pvList, err := p.host.GetKubeClient().Core().PersistentVolumes().List(v1.ListOptions{LabelSelector: labels.Everything().String()})
+	kubeClient := p.host.GetKubeClient()
+	if kubeClient == nil {
+		return fmt.Errorf("glusterfs: failed to get kube client when collecting gids")
+	}
+	pvList, err := kubeClient.Core().PersistentVolumes().List(v1.ListOptions{LabelSelector: labels.Everything().String()})
 	if err != nil {
 		glog.Errorf("glusterfs: failed to get existing persistent volumes")
 		return err
@@ -761,7 +769,11 @@ func (p *glusterfsVolumeProvisioner) createEndpointService(namespace string, epS
 			Ports:     []v1.EndpointPort{{Port: 1, Protocol: "TCP"}},
 		}},
 	}
-	_, err = p.plugin.host.GetKubeClient().Core().Endpoints(namespace).Create(endpoint)
+	kubeClient := p.plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return nil, nil, fmt.Errorf("glusterfs: failed to get kube client when creating endpoint service")
+	}
+	_, err = kubeClient.Core().Endpoints(namespace).Create(endpoint)
 	if err != nil && errors.IsAlreadyExists(err) {
 		glog.V(1).Infof("glusterfs: endpoint [%s] already exist in namespace [%s]", endpoint, namespace)
 		err = nil
@@ -781,7 +793,7 @@ func (p *glusterfsVolumeProvisioner) createEndpointService(namespace string, epS
 		Spec: v1.ServiceSpec{
 			Ports: []v1.ServicePort{
 				{Protocol: "TCP", Port: 1}}}}
-	_, err = p.plugin.host.GetKubeClient().Core().Services(namespace).Create(service)
+	_, err = kubeClient.Core().Services(namespace).Create(service)
 	if err != nil && errors.IsAlreadyExists(err) {
 		glog.V(1).Infof("glusterfs: service [%s] already exist in namespace [%s]", service, namespace)
 		err = nil
@@ -794,7 +806,11 @@ func (p *glusterfsVolumeProvisioner) createEndpointService(namespace string, epS
 }
 
 func (d *glusterfsVolumeDeleter) deleteEndpointService(namespace string, epServiceName string) (err error) {
-	err = d.plugin.host.GetKubeClient().Core().Services(namespace).Delete(epServiceName, nil)
+	kubeClient := d.plugin.host.GetKubeClient()
+	if kubeClient == nil {
+		return fmt.Errorf("glusterfs: failed to get kube client when deleting endpoint service")
+	}
+	err = kubeClient.Core().Services(namespace).Delete(epServiceName, nil)
 	if err != nil {
 		glog.Errorf("glusterfs: error deleting service %s/%s: %v", namespace, epServiceName, err)
 		return fmt.Errorf("error deleting service %s/%s: %v", namespace, epServiceName, err)

--- a/pkg/volume/util/util.go
+++ b/pkg/volume/util/util.go
@@ -149,6 +149,9 @@ func GetSecretForPV(secretNamespace, secretName, volumePluginName string, kubeCl
 }
 
 func GetClassForVolume(kubeClient clientset.Interface, pv *v1.PersistentVolume) (*storage.StorageClass, error) {
+	if kubeClient == nil {
+		return nil, fmt.Errorf("Cannot get kube client")
+	}
 	// TODO: replace with a real attribute after beta
 	className, found := pv.Annotations["volume.beta.kubernetes.io/storage-class"]
 	if !found {

